protocol-units/bridge/contracts/src/AtomicBridgeInitator.sol

```diff
@@ -0,0 +1,90 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.22;
+
+import {IAtomicBridgeInitiator} from "./IAtomicBridgeInitiator.sol";
+import {IWETH9} from "./IWETH9.sol";
+import {Initializable} from "@openzeppelin/contracts/proxy/utils/Initializable.sol";
+
+
+contract AtomicBridgeInitiator is IAtomicBridgeInitiator, Initializable {
+    enum MessageState { INITIALIZED, COMPLETED, REFUNDED }
+
+    struct BridgeTransfer {
+        uint256 amount;
+        address originator;
+        bytes32 recipient;
+        bytes32 hashLock;
+        uint256 timeLock; // in blocks
+        MessageState state;
+    }
+
+    mapping(bytes32 => BridgeTransfer) public bridgeTransfers;
+    IWETH9 public weth;
+    uint256 private nonce;
+
+    function initialize(address _weth) public initializer {
+        if (_weth == address(0)) {
+            revert ZeroAddress();
+        }
+        weth = IWETH9(_weth);
+    }
+
+    function initiateBridgeTransfer(uint256 wethAmount, bytes32 recipient, bytes32 hashLock, uint256 timeLock)
+        external
+        payable
+        returns (bytes32 bridgeTransferId)
+    {
+        address originator = msg.sender;
+        uint256 ethAmount = msg.value;
+        uint256 totalAmount = wethAmount + ethAmount;
+        // Ensure there is a valid total amount
+        if (totalAmount == 0) {
+            revert ZeroAmount();
+        }
+        // If msg.value is greater than 0, convert ETH to WETH
+        if (ethAmount > 0) weth.deposit{value: ethAmount}();
+        //Transfer WETH to this contract, revert if transfer fails
+        if (wethAmount > 0) {
+            if (!weth.transferFrom(originator, address(this), wethAmount)) revert WETHTransferFailed();
+        }
+
+        nonce++; //increment the nonce
+        bridgeTransferId =
+            keccak256(abi.encodePacked(originator, recipient, hashLock, timeLock, block.timestamp, nonce));
+
+        // Check if the bridge transfer already exists
+        if (bridgeTransfers[bridgeTransferId].amount != 0) revert BridgeTransferInvalid();
```

Comment:
Additionally, the error message currently does not accurately reflect the violation. It should instead be labeled as "BridgeTransferAlreadyExists" to provide a specific error that enables the Bridge to manage it effectively.

Reply:
we could check on timelock that's probably better. I think this was here in an earlier version ..

Reply:
Given the characteristics of the hashing function and nonce, it is impossible to generate a previously existing bridgeTransferId. Therefore, the corresponding statement can be omitted.

Reply:
yes no need for the extra compute. Every call here will generate a unique id.

Reply:
@cde8 if you access a non-existent resource, it will return default type values like 0, 0x0 or false.

e.g. a non-existent bridgeTransferId amount would return 0. Since bridge transfers must have over 0 amount, we can safely assume that if a bridgeTransferId has an amount different than 0, then it exists.